### PR TITLE
Home work 7

### PR DESCRIPTION
**Pull request by UnitOmega13**

### CODE REVIEW

File: src/Storage.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+class Storage<T> implements List<T> {
```
Comment (Sofasmile) [inline src/Storage.java]: let's use own List interface

maybe better use other class name, fo example ArrayList?

File: src/Storage.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    private T[] value;
+
+    Storage(){
+        this.value = (T[]) new Objects[0];
```
Comment (Sofasmile) [inline src/Storage.java]: by default constructors create an empty list with an initial capacity of ten.

File: src/Storage.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Do not have new index for operation");
+        }
+        return (T) value;
+    }
+
+    @Override
+    public void add(T valueOfElement) {
+        T[] tempArrayForValues = (T[]) value;
+        value = (T[]) new Object[tempArrayForValues.length + 1];
```
Comment (Sofasmile) [inline src/Storage.java]: we need increase size of the array in 1.5

File: src/Storage.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            System.arraycopy(value, index + 1, temp, index, value.length - index - 1);
+            value = temp;
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Do not have new index for operation");
+        }
+        return (T) value;
+    }
+
+    @Override
+    public void add(T valueOfElement) {
```
Comment (Sofasmile) [inline src/Storage.java]: try to use method add(T value, int index) to simplify code

File: src/Storage.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+class Storage<T> implements List<T> {
+    private T[] value;
```
Comment (Sofasmile) [inline src/Storage.java]: better change value -> values

File: src/Storage.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    private T removeArrayElement(int index) {
+        try {
+            T[] temp = (T[]) new Object[value.length - 1];
+            System.arraycopy(value, 0, temp, 0, index);
+            System.arraycopy(value, index + 1, temp, index, value.length - index - 1);
+            value = temp;
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Do not have new index for operation");
+        }
+        return (T) value;
```
Comment (Sofasmile) [inline src/Storage.java]: we need return only one element here

File: src/Storage.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Override
+    public T remove(int index) {
+        return removeArrayElement(index);
+    }
+
+    @Override
+    public T remove(T t) {
+        for (int i = 0; i < value.length; i++){
+            if (value[i] == t){
```
Comment (Sofasmile) [inline src/Storage.java]: test what happens if you write T = String

File: src/Storage.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                return removeArrayElement(i);
+            }
+        }
+        System.out.println("Element does not exist in Array");
```
Comment (Sofasmile) [inline src/Storage.java]: better create custom exception for this purpose

Comment (UnitOmega13) [reply src/Storage.java]: Changed everything according to comments

File: src/ListForTask.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    void set(T value, int index);
+
+    T remove(int index);
+
+    T remove(T t);
+
+    int size();
+
+    boolean isEmpty();
+}
```
Comment (Sofasmile) [inline src/ListForTask.java:19]: add a blank line after }

File: src/ListForTask.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+public interface ListForTask<T> {
+    void add(T value);
+
+    void add(T value, int index);
+
+    void addAll(ListForTask<T> listForTask);
```
Comment (Sofasmile) [inline src/ListForTask.java:6]: you can use standard List interface:
     void addAll(List<T> listForTask);

File: src/ArrayListForTaskStorage.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+class ArrayListForTaskStorage<T> implements ListForTask<T> {
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: rename class to ArrayList

File: src/ArrayListForTaskStorage.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        return values[index];
+    }
+
+    @Override
+    public void add(T valueOfElement) {
+        T[] tempArrayForValues = (T[]) values;
+        values = (T[]) new Object[tempArrayForValues.length + tempArrayForValues.length >> 1];
+        System.arraycopy(tempArrayForValues, 0, values, 0, tempArrayForValues.length);
+        values[values.length - 1] = valueOfElement;
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: if I understand this method will insert value in the last position in the array? //9

File: src/ArrayListForTaskStorage.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            System.arraycopy(values, index + 1, tempArrayList, index, values.length - index - 1);
+            values = tempArrayList;
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Do not have this index");
+        }
+        return values[index];
+    }
+
+    @Override
+    public void add(T valueOfElement) {
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: try use method add(T valueOfElement, int index) to simplify code

File: src/ArrayListForTaskStorage.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        } catch (NullPointerException e) {
+            System.out.println("ListForTask is null!");
+        }
+    }
+
+    @Override
+    public T get(int index) {
+        try {
+            return values[index];
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: list = new ArrayList();
list.add(1);
list.add(2);
list.get(4);
test what happens here?

File: src/ArrayListForTaskStorage.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Index is out of bound!");
+            return null;
+        }
+    }
+
+    @Override
+    public void set(T valueOfElement, int index) {
+        try {
+            values[index] = valueOfElement;
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: the same, try to test this method as method get

File: src/ArrayListForTaskStorage.java [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Override
+    public T remove(int index) {
+        return removeArrayElement(index);
+    }
+
+    @Override
+    public T remove(T t) {
+        try {
+            if (t instanceof String) {
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: what does this if statement do?

File: src/ArrayListForTaskStorage.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        } catch (ArrayIndexOutOfBoundsException e) {
+            System.out.println("Element does not exist in Array");
+        }
+        return t;
+    }
+
+    @Override
+    public int size() {
+        return values.length;
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: what is the size of empty list?

File: src/ArrayListForTaskStorage.java [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return values.length == 0;
```
Comment (Sofasmile) [inline src/ArrayListForTaskStorage.java]: what does return this method when list is empty?

File: src/ArrayList.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+	public ArrayList() {
+		this.newOwnArray = (T[]) new Object[CAPACITY_OF_ARRAY];
+	}
+
+	@Override
+	public void add(T value) {
+		if (arraySize >= newOwnArray.length - 1) {
+			newOwnArray = Arrays.copyOf(newOwnArray, newOwnArray.length + (newOwnArray.length >> 1));
+		}
+		newOwnArray[arraySize] = value;
```
Comment (Sofasmile) [inline src/ArrayList.java]: you can write it more simple:
newOwnArray[arraySize] = value;
		arraySize++;
   ->
newOwnArray[arraySize++] = value;

**Review by Sofasmile — APPROVED** [review]
good

**Pull request closed without merging**